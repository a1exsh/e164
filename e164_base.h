/*
 * E.164 Telephone Number Type for PostgreSQL: base functions
 *
 * Copyright (c) 2007-2011, Michael Glaesemann
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the
 * distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#ifndef E164_BASE_H
#define E164_BASE_H

#include "postgres.h"

#define E164_PREFIX '+'

typedef enum E164StructureLimit
{
    E164MaximumNumberOfDigits = 15,
    E164PrefixStringLength    = 1,      /* = strlen(E164_PREFIX) */
/*
 * E164MaxStringLength        = E164MaximumNumberOfDigits + E164PrefixStringLength
 * Note this does *not* include the string terminator
 */
    E164MaximumStringLength   = 16,
/*
 * E164MinimumStringLength is pretty conservative:
 * prefix (1) + country code (1) + subscriber number (1)
 */
    E164MinimumStringLength   = 3,
    E164MinimumNumberOfDigits = 2,

    E164MaximumCountryCodeLength               = 3,
    E164GeographicAreaMinimumCountryCodeLength = 1,
    E164GeographicAreaMaximumCountryCodeLength = 3,     /* = E164MaximumCountryCodeLength */
    E164GlobalServiceCountryCodeLength         = 3,     /* = E164MaximumCountryCodeLength */
    E164NetworkCountryCodeLength               = 3,     /* = E164MaximumCountryCodeLength */
    E164GroupOfCountriesCountryCodeLength      = 3,     /* = E164MaximumCountryCodeLength */

/*
 * Minimum Subscriber Number lengths for the various E164Types
 * These are absolute (and unrealistic) minimums. However, true
 * minimums are country specific, and until this implementation
 * is country-code specific, this should do.
 */
    E164GeographicAreaMinimumSubscriberNumberLength   = 1,
    E164GlobalServiceMinimumSubscriberNumberLength    = 1,
    E164NetworkMinimumSubscriberNumberLength          = 2,
    E164GroupOfCountriesMinimumSubscriberNumberLength = 2
} E164StructureLimit;

typedef enum E164ParseResult
{
    E164NoParseError = 0,
    E164ParseErrorBadFormat = 2,
    E164ParseErrorInvalidPrefix,
    E164ParseErrorStringTooLong,
    E164ParseErrorStringTooShort,
    E164ParseErrorInvalidType,
    E164ParseErrorNoSubscriberNumberDigits,
    E164ParseErrorUnassignedType,
    E164ParseErrorTypeLengthMismatch
} E164ParseResult;

typedef int4 E164CountryCode;
typedef uint64 E164;

/*
 * There are four types of assigned E164:
 *    * Geographic Area numbers
 *    * Global Service numbers
 *    * Network numbers
 *    * Group of Countries numbers
 * Each of these E164 number types have well-defined formats.
 *
 * There are three types of unassigned E164 as well:
 *     * Reserved numbers
 *     * Spare codes with notes
 *     * Spare codes without notes
 *
 * For the purposes of the implementation, all unassigned E164 numbers
 * will be considered invalid.
 *
 * An E164's type can be determined by inspecting its country code --
 * (at most) the first three digits. (Some Geographic Area country codes are one
 * or 2 digits in length.)
 *
 * Some country codes are reserved, and others are spare (as of yet unassigned).
 * These reserved and spare codes are rejected as invalid by this implementation,
 * in part because one cannot determine such a number's type and therefore whether
 * the number matches the format of its type.
 */
typedef enum E164Type
{
    E164GeographicArea,
    E164GlobalService,
    E164Network,
    E164GroupOfCountries,
    /* Unassigned codes */
    E164Reserved,
    E164SpareWithNote,
    E164SpareWithoutNote,
    E164Invalid
} E164Type;

/*
static const char * e164TypeName[] = {
    "E.164 Geographic Area",
    "E.164 Global Service",
    "E.164 Network",
    "E.164 Group of Countries",
    "E.164 Reserved",
    "E.164 Spare with Note",
    "E.164 Spare without Note",
    "E.164 Invalid"
};
*/

const E164Type e164TypeFor[1000];

extern E164ParseResult e164FromString (E164 * aNumber, const char * aString,
                                       E164CountryCode * aCode);
extern int stringFromE164 (char * aString, int stringLength, E164 aNumber);
extern int countryCodeStringFromE164 (char * aString, int stringLength,
                                      E164 aNumber);

extern int64 e164Comparison (E164 firstNumber, E164 secondNumber);

#endif /* !E164_BASE_H */
