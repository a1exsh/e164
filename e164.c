/*
 * E.164 Telephone Number Type for PostgreSQL: PG type interface
 *
 * Copyright (c) 2007-2011, Michael Glaesemann
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the
 * distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include "postgres.h"
#include "libpq/pqformat.h"
#include "access/hash.h"
#include "e164_base.h"

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

/*
 * makeText function taken from AndrewSN's ip4r
 */
static text *
makeText(int stringLength)
{
    text * textString = (text *) palloc0(stringLength + VARHDRSZ);
    SET_VARSIZE(textString, stringLength + VARHDRSZ);
    return textString;
}

static void
handleE164ParseError(E164ParseResult error, const char * string,
                     E164CountryCode countryCode)
{
    switch (error) {
        case E164ParseErrorBadFormat:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("invalid E164 number format: \"%s\"", string),
                     errhint("E164 numbers begin with a \"+\" followed by digits.")));
            break;
        case E164ParseErrorInvalidPrefix:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("invalid E164 prefix: \"%s\"", string),
                     errhint("E164 numbers must begin with \"%c\".", E164_PREFIX)));
            break;
        case E164ParseErrorStringTooLong:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("string too long: \"%s\"", string),
                     errhint("E164 values must have at most %d digits.", E164MaximumNumberOfDigits)));
            break;
        case E164ParseErrorStringTooShort:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("string too short \"%s\"", string),
                     errhint("E164 numbers must have at least %d digits.", E164MinimumNumberOfDigits)));
            break;
        case E164ParseErrorInvalidType:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("invalid E164 country code for E164 number \"%s\": %d", string, countryCode)));
            break;
        case E164ParseErrorNoSubscriberNumberDigits:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("no subscriber number digits in E164 number \"%s\"", string)));
            break;
        case E164ParseErrorUnassignedType:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("unassigned country code for E164 number \"%s\": %d", string, countryCode)));
            break;
        case E164ParseErrorTypeLengthMismatch:
            ereport(ERROR,
                    (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
                     errmsg("inconsistent length and country code for E164 number \"%s\" (country code: %d)", string, countryCode)));
            break;
        default:
            AssertArg(false);
    }
}

/*
 * PostgreSQL Interface functions
 */
#define DatumGetE164P(X) DatumGetInt64(X)
#define E164PGetDatum(X) Int64GetDatum(X)

#define PG_GETARG_E164(X) PG_GETARG_INT64((int64) X)
#define PG_RETURN_E164(X) PG_RETURN_INT64((int64) X)

Datum e164_in(PG_FUNCTION_ARGS);
Datum e164_out(PG_FUNCTION_ARGS);

Datum e164_hash(PG_FUNCTION_ARGS);
Datum e164_send(PG_FUNCTION_ARGS);
Datum e164_recv(PG_FUNCTION_ARGS);

Datum e164_lt(PG_FUNCTION_ARGS);
Datum e164_le(PG_FUNCTION_ARGS);
Datum e164_eq(PG_FUNCTION_ARGS);
Datum e164_ge(PG_FUNCTION_ARGS);
Datum e164_gt(PG_FUNCTION_ARGS);
Datum e164_ne(PG_FUNCTION_ARGS);

Datum e164_cmp(PG_FUNCTION_ARGS);

Datum e164_cast_to_text(PG_FUNCTION_ARGS);

Datum country_code(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(e164_in);
Datum
e164_in(PG_FUNCTION_ARGS)
{
    const char * theString = PG_GETARG_CSTRING(0);
    E164 theNumber;
    E164CountryCode theCode;
    E164ParseResult parseResult = e164FromString(&theNumber, theString, &theCode);
    if (E164NoParseError == parseResult)
    {
        PG_RETURN_E164(theNumber);
    }
    else
        handleE164ParseError(parseResult, theString, theCode);

    return 0;	/* keep compiler quiet */
}

PG_FUNCTION_INFO_V1(e164_out);
Datum
e164_out(PG_FUNCTION_ARGS)
{
    E164	theNumber = PG_GETARG_E164(0);
    char * theString = palloc(E164MaximumStringLength + 1);
    (void) stringFromE164(theString, E164MaximumStringLength + 1, theNumber);
    PG_RETURN_CSTRING(theString);
}

PG_FUNCTION_INFO_V1(e164_recv);
Datum
e164_recv(PG_FUNCTION_ARGS)
{
    StringInfo buf = (StringInfo) PG_GETARG_POINTER(0);
    PG_RETURN_E164((E164) pq_getmsgint(buf, sizeof(E164)));
}

PG_FUNCTION_INFO_V1(e164_send);
Datum
e164_send(PG_FUNCTION_ARGS)
{
    E164 arg1 = PG_GETARG_E164(0);
    StringInfoData buf;

    pq_begintypsend(&buf);
    pq_sendint(&buf, arg1, sizeof(E164));
    PG_RETURN_BYTEA_P(pq_endtypsend(&buf));
}

PG_FUNCTION_INFO_V1(e164_cast_to_text);
Datum
e164_cast_to_text(PG_FUNCTION_ARGS)
{
    E164	theNumber = PG_GETARG_E164(0);
    char	buffer[E164MaximumStringLength + 1];
    int		stringLength;
    text *	textString;

    stringLength = stringFromE164(buffer, E164MaximumStringLength + 1,
                                  theNumber);

    textString = makeText(stringLength);
    memcpy(VARDATA(textString), buffer, stringLength);

    PG_RETURN_TEXT_P(textString);
}

PG_FUNCTION_INFO_V1(country_code);
Datum
country_code(PG_FUNCTION_ARGS)
{
    E164	aNumber = PG_GETARG_E164(0);
    char	buffer[E164MaximumCountryCodeLength + 1];
    int		stringLength;
    text *	textString;

    stringLength = countryCodeStringFromE164(buffer,
                                             E164MaximumCountryCodeLength + 1,
                                             aNumber);

    textString = makeText(stringLength);
    memcpy(VARDATA(textString), buffer, stringLength);

    PG_RETURN_TEXT_P(textString);
}

PG_FUNCTION_INFO_V1(e164_lt);
Datum
e164_lt(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 > e164Comparison(PG_GETARG_E164(0),
                                      PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_le);
Datum
e164_le(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 >= e164Comparison(PG_GETARG_E164(0),
                                       PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_eq);
Datum
e164_eq(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 == e164Comparison(PG_GETARG_E164(0),
                                       PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_ge);
Datum
e164_ge(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 <= e164Comparison(PG_GETARG_E164(0),
                                       PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_gt);
Datum
e164_gt(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 < e164Comparison(PG_GETARG_E164(0),
                                      PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_ne);
Datum
e164_ne(PG_FUNCTION_ARGS)
{
    PG_RETURN_BOOL(0 != e164Comparison(PG_GETARG_E164(0),
                                       PG_GETARG_E164(1)));
}

PG_FUNCTION_INFO_V1(e164_cmp);
Datum
e164_cmp(PG_FUNCTION_ARGS)
{
    int64 comparison = e164Comparison(PG_GETARG_E164(0),
                                      PG_GETARG_E164(1));
    int result = ((comparison < 0) ?
                  -1 :
                  ((comparison > 0) ?
                   1 :
                   0));
    PG_RETURN_INT32(result);
}

PG_FUNCTION_INFO_V1(e164_hash);
Datum
e164_hash(PG_FUNCTION_ARGS)
{
    E164 arg1 = PG_GETARG_E164(0);
    return hash_any((unsigned char *)&arg1, sizeof(E164));
}
