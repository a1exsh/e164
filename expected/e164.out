-- E164 type regression test SQL script
SET search_path = public;
\set ECHO none
psql:e164.sql:17: NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "info_pkey" for table "info"
psql:e164.sql:17: NOTICE:  CREATE TABLE / UNIQUE will create implicit index "info_installed_at_key" for table "info"
psql:e164.sql:17: NOTICE:  CREATE TABLE / UNIQUE will create implicit index "info_scm_revision_key" for table "info"
psql:e164.sql:24: NOTICE:  type "e164" is not yet defined
DETAIL:  Creating a shell type definition.
psql:e164.sql:29: NOTICE:  argument type e164 is only a shell
psql:e164.sql:34: NOTICE:  return type e164 is only a shell
psql:e164.sql:39: NOTICE:  argument type e164 is only a shell
SET SEARCH_PATH to public, e164;
CREATE TABLE telephone_numbers
(
    telephone_number e164 PRIMARY KEY
);
NOTICE:  CREATE TABLE / PRIMARY KEY will create implicit index "telephone_numbers_pkey" for table "telephone_numbers"
COPY telephone_numbers (telephone_number) FROM STDIN;
SELECT *
FROM telephone_numbers
ORDER BY telephone_number;
 telephone_number 
------------------
 +12078652196
 +13032899913
 +16094926522
 +16158551760
 +17137292424
 +18007246269
 +18125224008
 +18162212045
 +18162796113
 +18887355977
 +19412583400
 +74956260391
 +74959808440
 +74959808441
 +78122326983
 +78122328250
 +78122328260
 +78123254044
 +2023366848
 +2034810166
 +9123692496
 +9126410553
 +9126552802
 +9128102290
 +20222686035
 +20222777000
 +20225912502
 +27114314068
 +27116223170
 +27117825226
 +27119491263
 +81185242051
 +81188327353
 +81188361268
 +81188364224
 +81762226757
 +81762412341
 +441223207072
 +441223208301
 +442070342900
 +442073779923
 +442085187347
 +442087782777
 +442089816811
 +448456032458
 +551150557460
 +552124527106
 +556233576049
 +557932173229
 +559132294022
 +35312121220
 +35318572979
(52 rows)

SELECT *
FROM telephone_numbers
ORDER BY CAST(telephone_number AS text);
 telephone_number 
------------------
 +12078652196
 +13032899913
 +16094926522
 +16158551760
 +17137292424
 +18007246269
 +18125224008
 +18162212045
 +18162796113
 +18887355977
 +19412583400
 +20222686035
 +20222777000
 +20225912502
 +2023366848
 +2034810166
 +27114314068
 +27116223170
 +27117825226
 +27119491263
 +35312121220
 +35318572979
 +441223207072
 +441223208301
 +442070342900
 +442073779923
 +442085187347
 +442087782777
 +442089816811
 +448456032458
 +551150557460
 +552124527106
 +556233576049
 +557932173229
 +559132294022
 +74956260391
 +74959808440
 +74959808441
 +78122326983
 +78122328250
 +78122328260
 +78123254044
 +81185242051
 +81188327353
 +81188361268
 +81188364224
 +81762226757
 +81762412341
 +9123692496
 +9126410553
 +9126552802
 +9128102290
(52 rows)

SELECT telephone_number
    , is_consistent(telephone_number)
FROM telephone_numbers
ORDER BY telephone_number;
 telephone_number | is_consistent 
------------------+---------------
 +12078652196     | t
 +13032899913     | t
 +16094926522     | t
 +16158551760     | t
 +17137292424     | t
 +18007246269     | t
 +18125224008     | t
 +18162212045     | t
 +18162796113     | t
 +18887355977     | t
 +19412583400     | t
 +74956260391     | t
 +74959808440     | t
 +74959808441     | t
 +78122326983     | t
 +78122328250     | t
 +78122328260     | t
 +78123254044     | t
 +2023366848      | t
 +2034810166      | t
 +9123692496      | t
 +9126410553      | t
 +9126552802      | t
 +9128102290      | t
 +20222686035     | t
 +20222777000     | t
 +20225912502     | t
 +27114314068     | t
 +27116223170     | t
 +27117825226     | t
 +27119491263     | t
 +81185242051     | t
 +81188327353     | t
 +81188361268     | t
 +81188364224     | t
 +81762226757     | t
 +81762412341     | t
 +441223207072    | t
 +441223208301    | t
 +442070342900    | t
 +442073779923    | t
 +442085187347    | t
 +442087782777    | t
 +442089816811    | t
 +448456032458    | t
 +551150557460    | t
 +552124527106    | t
 +556233576049    | t
 +557932173229    | t
 +559132294022    | t
 +35312121220     | t
 +35318572979     | t
(52 rows)

SELECT telephone_number
    , country_code(telephone_number)
FROM telephone_numbers
ORDER BY telephone_number;
 telephone_number | country_code 
------------------+--------------
 +12078652196     | 1
 +13032899913     | 1
 +16094926522     | 1
 +16158551760     | 1
 +17137292424     | 1
 +18007246269     | 1
 +18125224008     | 1
 +18162212045     | 1
 +18162796113     | 1
 +18887355977     | 1
 +19412583400     | 1
 +74956260391     | 7
 +74959808440     | 7
 +74959808441     | 7
 +78122326983     | 7
 +78122328250     | 7
 +78122328260     | 7
 +78123254044     | 7
 +2023366848      | 20
 +2034810166      | 20
 +9123692496      | 91
 +9126410553      | 91
 +9126552802      | 91
 +9128102290      | 91
 +20222686035     | 20
 +20222777000     | 20
 +20225912502     | 20
 +27114314068     | 27
 +27116223170     | 27
 +27117825226     | 27
 +27119491263     | 27
 +81185242051     | 81
 +81188327353     | 81
 +81188361268     | 81
 +81188364224     | 81
 +81762226757     | 81
 +81762412341     | 81
 +441223207072    | 44
 +441223208301    | 44
 +442070342900    | 44
 +442073779923    | 44
 +442085187347    | 44
 +442087782777    | 44
 +442089816811    | 44
 +448456032458    | 44
 +551150557460    | 55
 +552124527106    | 55
 +556233576049    | 55
 +557932173229    | 55
 +559132294022    | 55
 +35312121220     | 353
 +35318572979     | 353
(52 rows)

SELECT telephone_number < a_telephone_number as lt
    , telephone_number <= a_telephone_number as le
    , telephone_number = a_telephone_number as eq
    , telephone_number >= a_telephone_number as ge
    , telephone_number > a_telephone_number as gt
    , telephone_number <> a_telephone_number as ne
FROM telephone_numbers
CROSS JOIN (SELECT CAST('+442073779923' AS e164)) AS a(a_telephone_number)
ORDER BY telephone_number;
 lt | le | eq | ge | gt | ne 
----+----+----+----+----+----
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 f  | t  | t  | t  | t  | f
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
(52 rows)

SELECT telephone_number < a_telephone_number as lt
    , telephone_number <= a_telephone_number as le
    , telephone_number = a_telephone_number as eq
    , telephone_number >= a_telephone_number as ge
    , telephone_number > a_telephone_number as gt
    , telephone_number <> a_telephone_number as ne
FROM telephone_numbers
CROSS JOIN (SELECT CAST('+442073779923' AS e164)) AS a(a_telephone_number)
ORDER BY telephone_number;
 lt | le | eq | ge | gt | ne 
----+----+----+----+----+----
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 t  | t  | f  | f  | t  | t
 f  | t  | t  | t  | t  | f
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
 f  | f  | f  | t  | f  | t
(52 rows)

CREATE INDEX telephone_number_hash_idx
ON telephone_numbers USING hash
(telephone_number) ;
DROP TABLE telephone_numbers;
-- Invalid input strings
-- non-numeric characters in string
SELECT CAST('+1234567890a' AS e164);
ERROR:  invalid E164 number format: "+1234567890a"
HINT:  E164 numbers begin with a "+" followed by digits.
SELECT CAST('+a' AS e164);
ERROR:  string too short "+a"
HINT:  E164 numbers must have at least 2 digits.
-- leading or trailing whitespace
SELECT CAST(' +1234567890' AS e164);
ERROR:  invalid E164 prefix: " +1234567890"
HINT:  E164 numbers must begin with "+".
SELECT CAST('+1234567890 ' AS e164);
ERROR:  invalid E164 number format: "+1234567890 "
HINT:  E164 numbers begin with a "+" followed by digits.
SELECT CAST(' +1234567890 ' AS e164);
ERROR:  invalid E164 prefix: " +1234567890 "
HINT:  E164 numbers must begin with "+".
-- no prefix
SELECT CAST('1234567890' AS e164);
ERROR:  invalid E164 prefix: "1234567890"
HINT:  E164 numbers must begin with "+".
-- too short for e164
SELECT CAST('+1' AS e164);
ERROR:  string too short "+1"
HINT:  E164 numbers must have at least 2 digits.
-- too long for e164
SELECT CAST('+12345678901234567890' AS e164);
ERROR:  string too long: "+12345678901234567890"
HINT:  E164 values must have at most 15 digits.
-- unassigned country codes
-- 0 is reserved
SELECT CAST('+01234567890' AS e164);
ERROR:  unassigned country code for E164 number "+01234567890"
-- 280 is Spare with Note
SELECT CAST('+280123456' AS e164);
ERROR:  unassigned country code for E164 number "+280123456"
-- 295 is Spare without Note
SELECT CAST('+295123456' AS e164);
ERROR:  unassigned country code for E164 number "+295123456"
-- inconsistent length and country code (subscriber number minimum lengths)
-- Geographic Area and Global Service numbers require a subscriber number
-- of at least one digit. As *all* E164 numbers must have a subscriber
-- number of at least one digit, Geographic Area and Global Service numbers
-- without subscriber numbers will be invalid in general and not raise
-- this specific error.
-- Geographic Area: minimum 1 digit
-- '+1' should actually be caught as a minimum length error before this
SELECT CAST('+1' AS e164);
ERROR:  string too short "+1"
HINT:  E164 numbers must have at least 2 digits.
SELECT CAST('+11' AS e164);
 e164 
------
 +11
(1 row)

-- These should raise no subscriber number digit errors
SELECT CAST('+81' AS e164);
ERROR:  no subscriber number digits in E164 number "+81"
SELECT CAST('+811' AS e164); -- okay
 e164 
------
 +811
(1 row)

SELECT CAST('+220' AS e164);
ERROR:  no subscriber number digits in E164 number "+220"
SELECT CAST('+2201' AS e164); -- okay
 e164  
-------
 +2201
(1 row)

-- Global Service: minimum 1 digit
SELECT CAST('+800' AS e164);
ERROR:  no subscriber number digits in E164 number "+800"
SELECT CAST('+808' AS e164);
ERROR:  no subscriber number digits in E164 number "+808"
SELECT CAST('+8001' AS e164); -- okay
 e164  
-------
 +8001
(1 row)

SELECT CAST('+8081' AS e164); -- okay
 e164  
-------
 +8081
(1 row)

-- Inconsistent Group of Countries and Network numbers should raise
-- inconsistent E164 number
-- Group of Countries: minimum 2 digits
SELECT CAST('+3881' AS e164);
ERROR:  inconsistent length and country code: "+3881"
SELECT CAST('+38812' AS e164); -- okay
  e164  
--------
 +38812
(1 row)

SELECT CAST('+3882' AS e164);
ERROR:  inconsistent length and country code: "+3882"
SELECT CAST('+38822' AS e164); -- okay
  e164  
--------
 +38822
(1 row)

-- Network: minimum 2 digits
SELECT CAST('+8701' AS e164);
ERROR:  inconsistent length and country code: "+8701"
SELECT CAST('+87019' AS e164); -- okay
  e164  
--------
 +87019
(1 row)

SELECT CAST('+8711' AS e164);
ERROR:  inconsistent length and country code: "+8711"
SELECT CAST('+87119' AS e164); -- okay
  e164  
--------
 +87119
(1 row)

